[microvm] Add pause and resume functionality

LucasGdosR opened the pull request:
# Overview
This implements pausing and resuming via control-plane messages to the VMM via co-design with the kernel. The VMM writes to a kernel register. The kernel polls this register during timer interrupts, and exits the VM if flagged. It pauses execution and waits for a resume messages, which might be later sent by the VMM.

This does not implement pausing and resuming for hyperlight.

# Attention points:
- `DEFAULT_VMM_PAUSE_CMD` must not collide with other exit codes for the MicroVM;
- When pausing, the main thread sends to the memory thread, then receives from the vCPU thread. The main thread only receives from the vCPU thread after the vCPU thread has read the memory in the kernel that the memory thread wrote, so there's a total order in these events.;
- When resuming, the main thread sends to the memory thread, then receives from the memory thread, then sends to the vCPU thread. This means there's also a total order, and the vCPU thread will never resume the MicroVM with a pause request written in the kernel's memory.

ppenna commented:
@LucasGdosR does this superspeed #854?

LucasGdosR commented:
> @LucasGdosR does this superspeed #854?

Kind of. It's the same commit, I just switched the branch name to match the standard for regular PRs.

I'll do some manual testing before asking for a proper review.

ppenna commented:
Oh, I see. Thanks for being careful. The good news is that you can always push matching branches. The CI Will run only on non-draft PRs.

LucasGdosR commented on src/microvm/src/vmm/microvm/mod.rs:
```diff
@@ -554,9 +561,35 @@ impl Vmm {
+        self.microvm
+            .lock()
+            .map_err(|e| anyhow::anyhow!("failed to acquire lock {e:?}"))?
+            .async_pause()?;
```
I'm testing by hardcoding messages into the I/O thread (not in this commit). This call to `async_pause()` never happens, because `self.microvm` cannot be locked (although the error message never shows up). It seems to be locked elsewhere. The `vcpu_thread` calls:
```
microvm_clone
                .lock()
                .map_err(|e| anyhow::anyhow!("failed to acquire lock {e:?}"))?
                .run()
```
I'm thinking that's the problem. There shouldn't be a single lock that's used both for running the microvm and for pausing it, otherwise there's no way to interrupt it. Do you have suggestions on how to organize this into two different locks, or something else?

ppenna replied:
It is likely that some rework is required:

At a high level, to handle a snapshot you:
- Receive snapshot signal
- Sinal the kernel that a snapshot will happen
- The kernel issues a VM exit
- The vCPU unlocks the required resources

ppenna commented:
@LucasGdosR could you reopen #919 and have it independently merged with this one? We should have 1 PR for each feature/enhancement/bugfix whenever possible. As you stated in #919 the changed that you introduced there, but then later incorporated here, can be merged independently.

LucasGdosR commented:
I have reopened https://github.com/nanvix/nanvix/pull/919 and I'll include the `orchestrator.rs` part in it, without the pause-resume functionality. It's also duplicated code.

LucasGdosR committed:
[microvm] F: Implementing pause-resume functionality with kernel co-design

This implements pausing and resuming for the MicroVM only. It doesn't implement it for hyperlight yet.

In `mod.rs`, I implemented two closures that are called by the memory thread when it receives a request to pause. They write to the kernel's memory. Each of them holds a different Arc. I don't find it elegant, but hyperlight wouldn't have an Arc to pass, so I did it for compatibility with hyperlight. These closures write a magic value to a magic register. They are defined in `lib.rs`.

In the kernel, in `clock.rs`, the magic address is read and checked if it matches with the magic value. If it does, `::arch::io::out32()` causes a VM exit and sends another magic value that's also defined in `lib.rs`.

The magic value sent by the kernel is read by the Emulator in `emulator.rs`. It is passed to the MicroVM. Note this value that's passed must not collide with any existing exit code. It 0x3000 is already taken, then we need to change it in `lib.rs` to something else, or change the value when the Emulator passes it to the MicroVM.

In `microvm.rs`, the exit status is checked, and the vCPU thread informs the main thread that is has paused. Then, it loops, waiting for a message to resume. I added some timeout infrastructure, but I'm not sure it makes sense in this case. Resume messages might take arbitrarily long, and that's generally fine. I think timeouts should occur on the other end, the one sending the `Resume` message, rather than the one waiting for a `Resume` message.

Finally, `orchestrator.rs` changed because `VcpuControlResponse::Paused` is now constructed, so it lost its trailing underscore. Since I was already editing the file, I improved its trace messages.

As I mentioned, this does not implement pausing and resuming for hyperlight. To do that, we need to write similar closures / functions that write to the kernel's memory, and add similar behavior to the Emulator / MicroVM, where the exit context is checked and the vCPU thread interacts with the main thread via message passing. Resuming hyperlight might not be possible at this point, but pausing the program  and restarting it should be.

LucasGdosR commented:
@ppenna @csegarragonz this PR is now ready to be reviewed. I hope it'll be easy now that most of the code was merged in previous PRs. If there are no issues with it, I'll convert it to a regular PR.

I have tested the pause-resume functionality locally with a custom user program which you can find here, if you'd like https://github.com/nanvix/nanvix/tree/feature-long-running-rust-nostd-user-program

ppenna commented on src/microvm/src/vmm/microvm/microvm.rs:
```diff
@@ -319,7 +324,45 @@ impl MicroVm {
+                                    // Ok(VcpuControlCommand::Shutdown) => self.vcpu.poweroff(0),
+                                    Err(TryRecvError::Empty) => (),
+                                    Err(TryRecvError::Disconnected) => {
+                                        let reason: String = "the vmm has disconnected".to_string();
+                                        error!("run(): {reason}");
+                                        anyhow::bail!(reason)
```
This suggestion is incorrect.

ppenna commented on src/microvm/src/vmm/microvm/microvm.rs:
```diff
@@ -319,7 +324,45 @@ impl MicroVm {
+                                    // Ok(VcpuControlCommand::Shutdown) => self.vcpu.poweroff(0),
+                                    Err(TryRecvError::Empty) => (),
+                                    Err(TryRecvError::Disconnected) => {
+                                        let reason: String = "the vmm has disconnected".to_string();
+                                        error!("run(): {reason}");
+                                        anyhow::bail!(reason)
```
This suggestion is incorrect.

ppenna commented on src/microvm/src/vmm/microvm/microvm.rs:
```diff
@@ -319,7 +324,45 @@ impl MicroVm {
+
+                            let start: Instant = Instant::now();
+                            let mut counter: usize = 1;
+                            loop {
+                                match self.control_rx.try_recv() {
+                                    Ok(VcpuControlCommand::Resume) => break,
+                                    // NOTE: Should we add an option for shutting down? Like so:
+                                    // Ok(VcpuControlCommand::Shutdown) => self.vcpu.poweroff(0),
+                                    Err(TryRecvError::Empty) => (),
+                                    Err(TryRecvError::Disconnected) => {
+                                        let reason: String = "the vmm has disconnected".to_string();
+                                        error!("run(): {reason}");
+                                        anyhow::bail!(reason)
+                                    },
+                                }
+
+                                // NOTE: is it desirable to check for timeout in this case? If it is,
+                                // should we use a larger constant, considering snapshots might take long?
+
+                                // Log a warning and increment the counter every TIMEOUT_WARNING_INTERVAL_IN_MS ms.
+                                let elapsed_time: usize = start.elapsed().as_millis() as usize;
+                                if elapsed_time > TIMEOUT_WARNING_INTERVAL_IN_MS * counter {
+                                    warn!(
+                                        "{}ms have passed waiting for `ResumeMicroVm` message",
+                                        TIMEOUT_WARNING_INTERVAL_IN_MS * counter
+                                    );
+                                    counter += 1;
```
Could you link the issue of the exponential back-off timeout here?
